implementations/rust/ockam/ockam_api/src/nodes/service/tcp_inlets/session_replacer/mod.rs

```diff
@@ -0,0 +1,449 @@
+            Err(_) => {
+                warn!(outlet_multiaddresses = %self.parameters.outlet_address_selector, "timeout creating new tcp inlet");
+                Err(ApiError::core("timeout"))
+            }
+            Ok(Err(e)) => {
+                warn!(outlet_multiaddresses = %self.parameters.outlet_address_selector, err = %e, "failed to create tcp inlet");
+                Err(e)
+            }
+            Ok(Ok(route)) => Ok(route),
+        }
+    }
+
+    async fn close(&mut self) {
+        if let Some(status) = self.status.take() {
+            self.inlet.remove_route(&status.last_route_key);
+
+            let node_manager = if let Some(node_manager) = self.node_manager.upgrade() {
+                node_manager
+            } else {
+                warn!("An inlet close was issued after the NodeManager shut down, skipping.");
+                return;
+            };
+
+            let result = status.connection.close(&self.context, &node_manager);
+            if let Err(err) = result {
+                error!(?err, "Failed to close connection");
+            }
+        }
+    }
+
+    async fn on_session_down(&self) {
+        if let Some(status) = &self.status {
+            self.parameters
+                .terminal_notifier
+                .on_session_down(&status.original_multiaddr)
+                .await;
+        }
+    }
+
+    async fn on_session_replaced(&self) {
+        if let Some(status) = &self.status {
+            self.parameters
+                .terminal_notifier
+                .on_session_replaced(&status.original_multiaddr)
+                .await;
+        }
+    }
+}
+
+#[async_trait]
+impl AdditionalSessionReplacer for InletSessionReplacer {
+    async fn create_additional(&mut self) -> Result<Route> {
+        let node_manager = if let Some(node_manager) = self.node_manager.upgrade() {
+            node_manager
+        } else {
+            return Err(Error::new(
+                Origin::Node,
+                Kind::Cancelled,
+                "Node manager is dropped. Can't start UDP puncture for an Inlet.",
+            ));
+        };
+
+        let udp_transport = self
+            .udp_transport
+            .as_ref()
+            .ok_or_else(|| {
+                Error::new(
+                    Origin::Node,
+                    Kind::Invalid,
+                    "Couldn't create inlet with puncture",
+                )
+            })?
+            .clone();
+
+        let status = if let Some(status) = &mut self.status {
+            status
+        } else {
+            return Err(Error::new(
+                Origin::Api,
+                Kind::Internal,
+                "Error while creating additional session. Connection is absent",
+            ));
+        };
+
+        let transport_route = status.connection.transport_route();
+        let additional_sc_route = transport_route.clone() + DefaultAddress::SECURE_CHANNEL_LISTENER;
```

Comment:
Not the most common use case, but we are limiting this feature to the default secure channel listener. Should this be passed down from a higher layer? Or at least I'd add a TODO if it's important enough. Same with the other `DefaultAddress:*` usages below

Reply:
Added a TODO to extract it from the original MultiAddr